sql/group.go

```diff
@@ -0,0 +1,443 @@
+var _ aggregateImpl = &minAggregate{}
+var _ aggregateImpl = &sumAggregate{}
+
+type avgAggregate struct {
+	sumAggregate
+	count int
+}
+
+func (a *avgAggregate) New() aggregateImpl {
+	return &avgAggregate{}
+}
+
+func (a *avgAggregate) Add(datum parser.Datum) error {
+	if datum == parser.DNull {
+		return nil
+	}
+	if err := a.sumAggregate.Add(datum); err != nil {
+		return err
+	}
+	a.count++
+	return nil
+}
+
+func (a *avgAggregate) Result() (parser.Datum, error) {
+	sum, err := a.sumAggregate.Result()
+	if err != nil {
+		return parser.DNull, err
+	}
+	if sum == parser.DNull {
+		return sum, nil
+	}
+	switch t := sum.(type) {
+	case parser.DInt:
+		return parser.DFloat(t) / parser.DFloat(a.count), nil
+	case parser.DFloat:
+		return t / parser.DFloat(a.count), nil
+	default:
+		return parser.DNull, fmt.Errorf("unexpected SUM result type: %s", t.Type())
+	}
+}
+
+type countAggregate struct {
+	count int
+}
+
+func (a *countAggregate) New() aggregateImpl {
+	return &countAggregate{}
+}
+
+func (a *countAggregate) Add(datum parser.Datum) error {
+	if datum == parser.DNull {
+		return nil
+	}
+	switch t := datum.(type) {
+	case parser.DTuple:
+		for _, d := range t {
+			if d != parser.DNull {
+				a.count++
+				break
+			}
+		}
+	default:
+		a.count++
+	}
+	return nil
+}
+
+func (a *countAggregate) Result() (parser.Datum, error) {
+	return parser.DInt(a.count), nil
+}
+
+type maxAggregate struct {
+	max parser.Datum
+}
+
+func (a *maxAggregate) New() aggregateImpl {
+	return &maxAggregate{}
+}
+
+func (a *maxAggregate) Add(datum parser.Datum) error {
+	if datum == parser.DNull {
+		return nil
+	}
+	if a.max == nil {
+		a.max = datum
+		return nil
+	}
+	c := a.max.Compare(datum)
+	if c < 0 {
+		a.max = datum
+	}
+	return nil
+}
+
+func (a *maxAggregate) Result() (parser.Datum, error) {
+	if a.max == nil {
+		return parser.DNull, nil
+	}
+	return a.max, nil
+}
+
+type minAggregate struct {
+	min parser.Datum
+}
+
+func (a *minAggregate) New() aggregateImpl {
+	return &minAggregate{}
+}
+
+func (a *minAggregate) Add(datum parser.Datum) error {
+	if datum == parser.DNull {
+		return nil
+	}
+	if a.min == nil {
+		a.min = datum
+		return nil
+	}
+	c := a.min.Compare(datum)
+	if c > 0 {
+		a.min = datum
+	}
+	return nil
+}
+
+func (a *minAggregate) Result() (parser.Datum, error) {
+	if a.min == nil {
+		return parser.DNull, nil
+	}
+	return a.min, nil
+}
+
+type sumAggregate struct {
+	sum parser.Datum
+}
+
+func (a *sumAggregate) New() aggregateImpl {
+	return &sumAggregate{}
+}
+
+func (a *sumAggregate) Add(datum parser.Datum) error {
+	if datum == parser.DNull {
+		return nil
+	}
+	if a.sum == nil {
+		a.sum = datum
+		return nil
+	}
+
+	switch t := datum.(type) {
+	case parser.DInt:
+		if v, ok := a.sum.(parser.DInt); ok {
+			a.sum = v + t
+			return nil
+		}
+
+	case parser.DFloat:
+		if v, ok := a.sum.(parser.DFloat); ok {
+			a.sum = v + t
+			return nil
+		}
+	}
+
+	return fmt.Errorf("unexpected SUM argument type: %s", datum.Type())
```

Comment:
Yes, but this code is slightly different. Notice that the switch cases fall through on unexpected types.